backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/EnumClassLowering.kt

```diff
@@ -158,9 +158,88 @@ internal class EnumUsageLowering(val context: Context)
     }
 }
 
+// Look for when-constructs where subject is enum entry.
+// Replace branches that are comparisons with compile-time known enum entries
+// with comparisons of ordinals.
+internal class EnumWhenLowering(
+        private val context: Context,
+        private val enumEntryOrdinals: Map<ClassDescriptor, Int>)
+    : IrElementTransformerVoid(), FileLoweringPass {
+
+    override fun lower(irFile: IrFile) {
+        visitFile(irFile)
+    }
+
+    override fun visitBlock(expression: IrBlock): IrExpression {
+        if (!shouldLower(expression)) {
+            return super.visitBlock(expression)
+        }
+        // Will be initialized only when we found a branch that compares
+        // subject with compile-time known enum entry
+        val ordinalVariable: IrVariable by lazy {
+            val variable = createOrdinalVariable(expression)
+            expression.statements.add(1, variable)
+            variable
+        }
+        val areEqualByValue = context.ir.symbols.areEqualByValue.first {
+            it.owner.valueParameters[0].type == context.builtIns.intType
+        }
+        val whenExpr = expression.statements[1] as IrWhen
+        whenExpr.branches.filter { it.condition is IrCall }.forEach {
+            val eqEqCall = it.condition as IrCall
+            val entry = eqEqCall.getArguments()[1].second as? IrGetEnumValue
+            if (entry != null) {
+                val entryOrdinal = enumEntryOrdinals[entry.descriptor]!!
```

Comment:
Let's hold that until IR types will be merged, this code is IR-descriptors specific.